src/multiplex/client.rs

```diff
@@ -138,21 +238,131 @@ where
     }
 }
 
-// ===== ClientInner =====
+// ===== Pending =====
 
-struct Pending<Tag, Item> {
+/// A pending response used to track in-flight requests.
+///
+/// Each pending response contains an associated `Tag` that is provided
+/// from the `TagStore` used to uniquely identify a request/response pair.
+pub struct Pending<Tag, Response> {
     tag: Tag,
-    tx: tokio::sync::oneshot::Sender<ClientResponse<Item>>,
+    tx: tokio::sync::oneshot::Sender<ClientResponse<Response>>,
     span: tracing::Span,
 }
 
+impl<Tag, Response> Pending<Tag, Response> {
+    /// Get the tag associated with this pending request.
+    pub fn tag(&self) -> &Tag {
+        &self.tag
+    }
+
+    /// Get the span associated with this pending request.
+    pub fn span(&self) -> &tracing::Span {
+        &self.span
+    }
+}
+
+impl<Tag, Response> fmt::Debug for Pending<Tag, Response>
+where
+    Tag: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Pending")
+            .field("tag", &self.tag)
+            .field("span", &self.span)
+            .finish()
+    }
+}
+
+// ===== Builder =====
+
+/// Builder for [`Client`] this is used to configure the transport, pending store
+/// and service_handler.
+///
+/// # Defaults
+///
+/// By default this builder only requires a transport and sets a default pending store and
+/// error handler. The default pending store is just a [`VecDeque`] and the default
+/// error handler is just an empty closure.
+pub struct Builder<T, E, Request, P = VecDequePendingStore<T, Request>> {
+    transport: T,
+    on_service_error: Box<dyn FnOnce(E) + Send>,
+    pending_store: P,
+    _pd: PhantomData<fn(Request, E)>,
+}
+
+impl<T, E, Request, P> Builder<T, E, Request, P>
+where
+    T: Sink<Request> + TryStream + TagStore<Request, <T as TryStream>::Ok> + Send + 'static,
+    P: PendingStore<T, Request> + Send + 'static,
+    E: From<Error<T, Request>>,
+    E: 'static + Send,
+    Request: 'static + Send,
+    T::Ok: 'static + Send,
+    T::Tag: Send,
+{
+    fn new(transport: T) -> Builder<T, E, Request, VecDequePendingStore<T, Request>> {
+        Builder {
+            transport,
+            on_service_error: Box::new(|_| {}),
+            pending_store: VecDequePendingStore::default(),
+            _pd: PhantomData,
+        }
+    }
+
+    /// Set the provided pending store.
+    pub fn pending_store<P2>(self, pending_store: P2) -> Builder<T, E, Request, P2> {
```

Comment:
If we're willing to have this be generic, why not have the `on_service_error` be generic too? It's a builder after all.

Reply:
Ok I updated it to be that. I don't know of a better way for the default one than to box but I don't think its the end of the world to go this way.